File: Platforms/Win32/Reign.Video.D3D9/Shaders/PixelShader.cs
using System;
using Reign.Core;
using Reign_Video_D3D9_Component;

namespace Reign.Video.D3D9
{
	public class PixelShader : ShaderModel
	{
		#region Properties
		private PixelShaderCom pixelShaderCom;
		#endregion

		#region Constructors
		public PixelShader(DisposableI parent, string code, ShaderVersions shaderVersion)
		: base(parent, code, shaderVersion, ShaderTypes.PS)
		{
			try
			{
				var video = parent.FindParentOrSelfWithException<Video>();

				pixelShaderCom = new PixelShaderCom();
				var error = pixelShaderCom.Init(video.com, com);

				switch (error)
				{
					case PixelShaderErrors.PixelShader: Debug.ThrowError("PixelShader", "Failed to create pixel shader"); break;
				}
			}
			catch (Exception e)
			{
				Dispose();
				throw e;
			}
		}

		public override void Dispose()
		{
			disposeChilderen();
			if (pixelShaderCom != null)
			{
				pixelShaderCom.Dispose();
				pixelShaderCom = null;
			}
			base.Dispose();
		}
		#endregion

		#region Methods
		public void Apply()
		{
			pixelShaderCom.Apply();
		}
		#endregion
	}
}

File: Platforms/Win32/Reign_Video_D3D9_Component/PixelShader.cpp
#include "pch.h"
#include "PixelShader.h"

namespace Reign_Video_D3D9_Component
{
	#pragma region Constructors
	PixelShaderErrors PixelShaderCom::Init(VideoCom^ video, ShaderModelCom^ shaderModel)
	{
		null();
		this->video = video;

		IDirect3DPixelShader9* shaderTEMP = 0;
		// TODO: load data file file
		/*if (FAILED(video->device->CreatePixelShader((DWORD*)shaderModel->code->GetBufferPointer(), &shaderTEMP)))
		{
			return PixelShaderErrors::PixelShader;
		}*/
		this->shader = shaderTEMP;

		return PixelShaderErrors::None;
	}

	PixelShaderCom::~PixelShaderCom()
	{
		if (shader) shader->Release();
		null();
	}

	void PixelShaderCom::null()
	{
		shader = 0;
	}
	#pragma endregion

	#pragma region Methods
	void PixelShaderCom::Apply()
	{
		video->device->SetPixelShader(shader);
	}
	#pragma endregion
}